File: 20-concurrency/main.go
package main

import (
	"fmt"
	"sync"
	"time"
)

// ============================================================
// 示例 1: goroutine 基本使用
// ============================================================

func basicGoroutineDemo() {
	fmt.Println("=== goroutine 基本使用 ===")

	var wg sync.WaitGroup

	for i := 0; i < 4; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			fmt.Printf("goroutine %d 开始工作\n", id)
			time.Sleep(100 * time.Millisecond)
			fmt.Printf("goroutine %d 完成工作\n", id)
		}(i)
	}

	wg.Wait()
	fmt.Println("所有 goroutine 已完成")
	fmt.Println()
}

// ============================================================
// 示例 2: Channel 通信
// ============================================================

func channelDemo() {
	fmt.Println("=== Channel 通信 ===")

	// 无缓冲 channel
	ch := make(chan string)

	go func() {
		ch <- "你好，来自 goroutine！"
	}()

	msg := <-ch
	fmt.Println(msg)

	// 带缓冲 channel
	buffered := make(chan int, 3)
	buffered <- 1
	buffered <- 2
	buffered <- 3

	fmt.Println(<-buffered) // 1
	fmt.Println(<-buffered) // 2
	fmt.Println(<-buffered) // 3
	fmt.Println()
}

// ============================================================
// 示例 3: Select 多路复用
// ============================================================

func selectDemo() {
	fmt.Println("=== Select 多路复用 ===")

	ch1 := make(chan string)
	ch2 := make(chan string)

	go func() {
		time.Sleep(50 * time.Millisecond)
		ch1 <- "来自 channel 1"
	}()

	go func() {
		time.Sleep(100 * time.Millisecond)
		ch2 <- "来自 channel 2"
	}()

	// 等待两个 channel
	for i := 0; i < 2; i++ {
		select {
		case msg := <-ch1:
			fmt.Println(msg)
		case msg := <-ch2:
			fmt.Println(msg)
		}
	}
	fmt.Println()
}

// ============================================================
// 示例 4: Mutex 保护共享数据
// ============================================================

type SafeCounter struct {
	mu    sync.Mutex
	count int64
}

func (c *SafeCounter) Increment() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.count++
}

func (c *SafeCounter) GetCount() int64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.count
}

func mutexDemo() {
	fmt.Println("=== Mutex 保护共享数据 ===")

	counter := &SafeCounter{}
	var wg sync.WaitGroup

	for i := 0; i < 4; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 1000; j++ {
				counter.Increment()
			}
		}()
	}

	wg.Wait()
	fmt.Printf("最终计数: %d (期望 4000)\n", counter.GetCount())
	fmt.Println()
}

// ============================================================
// 示例 5: 生产者-消费者模式 (用 channel 实现)
// ============================================================

func producerConsumerDemo() {
	fmt.Println("=== 生产者-消费者 (channel) ===")

	ch := make(chan int, 8) // 带缓冲的 channel

	// 生产者
	go func() {
		for i := 0; i < 10; i++ {
			ch <- i
			fmt.Printf("生产: %d\n", i)
		}
		close(ch) // 关闭 channel 通知消费者
	}()

	// 消费者：range 自动在 channel 关闭后退出
	for val := range ch {
		fmt.Printf("消费: %d\n", val)
	}

	fmt.Println("生产者-消费者完成")
	fmt.Println()
}

// ============================================================
// 示例 6: Worker Pool 模式
// ============================================================

func workerPoolDemo() {
	fmt.Println("=== Worker Pool ===")

	jobs := make(chan int, 8)
	var wg sync.WaitGroup

	// 启动 4 个 worker
	for w := 0; w < 4; w++ {
		wg.Add(1)
		go func(workerID int) {
			defer wg.Done()
			for job := range jobs {
				fmt.Printf("Worker %d 处理任务 %d\n", workerID, job)
				time.Sleep(50 * time.Millisecond)
			}
		}(w)
	}

	// 提交 8 个任务
	for j := 0; j < 8; j++ {
		jobs <- j
	}
	close(jobs)

	wg.Wait()
	fmt.Println("所有任务完成")
	fmt.Println()
}

func main() {
	basicGoroutineDemo()
	channelDemo()
	selectDemo()
	mutexDemo()
	producerConsumerDemo()
	workerPoolDemo()
}

File: 20-concurrency/main.rs
use std::sync::mpsc;
use std::sync::{Arc, Condvar, Mutex};
use std::thread;
use std::time::Duration;

// ============================================================
// 示例 1: 基本线程创建
// ============================================================

fn basic_thread_demo() {
    println!("=== 基本线程创建 ===");

    let mut handles = vec![];

    for i in 0..4 {
        let handle = thread::spawn(move || {
            println!("线程 {} 开始工作", i);
            thread::sleep(Duration::from_millis(100));
            println!("线程 {} 完成工作", i);
        });
        handles.push(handle);
    }

    for handle in handles {
        handle.join().unwrap();
    }
    println!("所有线程已完成\n");
}

// ============================================================
// 示例 2: Channel 通信 (mpsc)
// ============================================================

fn channel_demo() {
    println!("=== Channel (mpsc) ===");

    // mpsc: multiple producer, single consumer
    let (tx, rx) = mpsc::channel();

    thread::spawn(move || {
        tx.send("你好，来自子线程！".to_string()).unwrap();
    });

    let msg = rx.recv().unwrap();
    println!("{}", msg);

    // 多个生产者
    let (tx, rx) = mpsc::channel();
    for i in 0..4 {
        let tx_clone = tx.clone();
        thread::spawn(move || {
            tx_clone.send(format!("消息来自线程 {}", i)).unwrap();
        });
    }
    drop(tx); // 丢弃原始发送端

    // 接收所有消息
    for msg in rx {
        println!("{}", msg);
    }
    println!();
}

// ============================================================
// 示例 3: Mutex 保护共享数据
// ============================================================

fn mutex_demo() {
    println!("=== Arc<Mutex<T>> 共享数据 ===");

    // Arc: 原子引用计数（跨线程共享所有权）
    // Mutex: 互斥锁
    let counter = Arc::new(Mutex::new(0i64));
    let mut handles = vec![];

    for _ in 0..4 {
        let counter = Arc::clone(&counter);
        let handle = thread::spawn(move || {
            for _ in 0..1000 {
                let mut count = counter.lock().unwrap();
                *count += 1;
            }
        });
        handles.push(handle);
    }

    for handle in handles {
        handle.join().unwrap();
    }

    println!("最终计数: {} (期望 4000)\n", *counter.lock().unwrap());
}

// ============================================================
// 示例 4: 条件变量 (生产者-消费者)
// ============================================================

fn condvar_demo() {
    println!("=== 条件变量 (生产者-消费者) ===");

    let queue = Arc::new((Mutex::new(Vec::<i32>::new()), Condvar::new()));
    let done = Arc::new(Mutex::new(false));

    // 生产者
    let queue_p = Arc::clone(&queue);
    let done_p = Arc::clone(&done);
    let producer = thread::spawn(move || {
        let (lock, cvar) = &*queue_p;
        for i in 0..10 {
            {
                let mut q = lock.lock().unwrap();
                q.push(i);
                println!("生产: {}", i);
            }
            cvar.notify_one();
        }
        // 标记完成
        *done_p.lock().unwrap() = true;
        cvar.notify_one();
    });

    // 消费者
    let queue_c = Arc::clone(&queue);
    let done_c = Arc::clone(&done);
    let consumer = thread::spawn(move || {
        let (lock, cvar) = &*queue_c;
        let mut consumed = 0;
        while consumed < 10 {
            let mut q = lock.lock().unwrap();
            while q.is_empty() {
                if *done_c.lock().unwrap() && q.is_empty() {
                    return;
                }
                q = cvar.wait(q).unwrap();
            }
            while let Some(val) = q.pop() {
                println!("消费: {}", val);
                consumed += 1;
            }
        }
    });

    producer.join().unwrap();
    consumer.join().unwrap();
    println!("生产者-消费者完成\n");
}

// ============================================================
// 示例 5: 简单线程池 (用 channel 实现)
// ============================================================

fn thread_pool_demo() {
    println!("=== 简单线程池 ===");

    let (tx, rx) = mpsc::channel::<Box<dyn FnOnce() + Send>>();
    let rx = Arc::new(Mutex::new(rx));
    let mut workers = vec![];

    // 启动 4 个 worker 线程
    for id in 0..4 {
        let rx = Arc::clone(&rx);
        let handle = thread::spawn(move || {
            loop {
                let task = {
                    let rx = rx.lock().unwrap();
                    rx.recv()
                };
                match task {
                    Ok(task) => {
                        println!("Worker {} 执行任务", id);
                        task();
                    }
                    Err(_) => break, // channel 关闭
                }
            }
        });
        workers.push(handle);
    }

    // 提交 8 个任务
    for i in 0..8 {
        tx.send(Box::new(move || {
            println!("  任务 {} 完成", i);
            thread::sleep(Duration::from_millis(50));
        }))
        .unwrap();
    }

    drop(tx); // 关闭 channel，worker 会退出循环

    for w in workers {
        w.join().unwrap();
    }
    println!("所有任务完成\n");
}

fn main() {
    basic_thread_demo();
    channel_demo();
    mutex_demo();
    condvar_demo();
    thread_pool_demo();
}
